presto-spark-base/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java

```diff
@@ -15,283 +15,367 @@
+    }
 
-            for (ScheduledSplit split : scheduledSplits) {
-                int taskId = Objects.hash(split.getPlanNodeId(), split.getSequenceId()) % numTasks;
-                if (taskId < 0) {
-                    taskId += numTasks;
-                }
+    private JavaPairRDD<Integer, PrestoSparkRow> createSourceRdd(
+            JavaSparkContext sparkContext,
+            Session session,
+            PlanFragment fragment,
+            PrestoSparkTaskExecutorFactoryProvider executorFactoryProvider,
+            CollectionAccumulator<SerializedTaskStats> taskStatsCollector,
+            TableWriteInfo tableWriteInfo)
+    {
+        // TODO: Possible in case of a broadcast join
+        checkArgument(fragment.getRemoteSourceNodes().isEmpty(), "source task with remote sources is not supported");
+
+        List<TableScanNode> tableScans = findTableScanNodes(fragment.getRoot());
+        checkArgument(
+                tableScans.size() == 1,
+                "exactly one table scan is expected in SOURCE_DISTRIBUTION fragment. fragmentId: %s, actual number of table scans: %s",
+                fragment.getId(),
+                tableScans.size());
+        verify(tableScans.size() == fragment.getTableScanSchedulingOrder().size());
+
+        TableScanNode tableScan = tableScans.get(0);
+
+        List<ScheduledSplit> splits = getSplits(session, tableScan);
+        shuffle(splits);
+        int initialPartitionCount = getSparkInitialPartitionCount(session);
+        int numTasks = Math.min(splits.size(), initialPartitionCount);
+        if (numTasks == 0) {
+            return JavaPairRDD.fromJavaRDD(sparkContext.emptyRDD());
+        }
 
-                assignedSplits.get(taskId).add(split);
-            }
+        List<List<ScheduledSplit>> assignedSplits = assignSplitsToTasks(splits, numTasks);
 
-            return assignedSplits;
+        // let the garbage collector reclaim the memory used by the decoded splits as soon as the task descriptor is encoded
+        splits = null;
+
+        ImmutableList.Builder<SerializedPrestoSparkTaskDescriptor> serializedTaskDescriptors = ImmutableList.builder();
+        for (int i = 0; i < assignedSplits.size(); i++) {
+            List<ScheduledSplit> splitBatch = assignedSplits.get(i);
+            PrestoSparkTaskDescriptor taskDescriptor = createSourceTaskDescriptor(session, tableWriteInfo, fragment, splitBatch);
+            // TODO: consider more efficient serialization or apply compression to save precious memory on the Driver
+            byte[] jsonSerializedTaskDescriptor = taskDescriptorJsonCodec.toJsonBytes(taskDescriptor);
+            serializedTaskDescriptors.add(new SerializedPrestoSparkTaskDescriptor(jsonSerializedTaskDescriptor));
+            // let the garbage collector reclaim the memory used by the decoded splits as soon as the task descriptor is encoded
+            assignedSplits.set(i, null);
```

Comment:
As a TODO, I think someday we want to have dedicated class for `assignedSplits`; just use a `Map<Integer, List<Split>>` and remove the map entry would also be better than set some element in list as null XDDD.

Reply:
I wouldn't try to overthink it. Although I agree that a List of Lists is not very readable, the scope of this variable is very low.

Reply:
@arhimondr : Setting some element to be null in a List to make garbage collector re-claim definitely works, but I don't really think it's a common way... so I would argue it's not a "overthink" :). The operation really need here is `remove`.

However, I understand we want to use null setting to keep the commit straightforward  for now. And it's too early to sweat on how to organize this part of the code for now. So I suggest to leave a comment.

Reply:
> Setting some element to be null in a List to make garbage collector re-claim definitely works, but I don't really think it's a common way... so I would argue it's not a "overthink" :).

Oh, i see what you mean. 

> So I suggest to leave a comment.

I have a comment there:

> // let the garbage collector reclaim the memory us

Do you think it requires further clarification?

Reply:
@arhimondr : I am just thinking we probably want to refactor the code in the future, even a `Map<Integer, List<Split>>` seems to be better than `List<List<Split>>` since now you can do `Map.remove(i)` instead of `List#set(i, null)` :)

I mean, to put an comment like the following:
```
// let the garbage collector reclaim the memory us
// TODO: Investigate if we can avoid this by not using List
```

Reply:
`List` also has a `remove(index)` method: https://docs.oracle.com/javase/8/docs/api/java/util/List.html#remove-int-

I just thought that an explicit nullification can be more readable, as I had an impression that's usually the way these tricks are implemented. But I don't have any strong opinion here.

I also don't think using `List<List>` is very exotic. It is what the `Lists<List>` is returned by `Lists.partition` from Guava. And we use it in Presto in several places to split a larger list into smaller partitions (batches).

Wrapping it into a class can make it a little bit more readable, but given the very small scope of this variable It feels like it is more of a matter of taste.

Reply:
> I just thought that an explicit nullification can be more readable, as I had an impression that's usually the way these tricks are implemented. But I don't have any strong opinion here.

Right, but removing front element in `List` is `O(n)`. I thought that's why you nullify it instead of remove it. We can also use `ArrayDeque`, but anyway, I agree it's just a matter of taste. Let's just keep it as is.

Reply:
Oh, right. It has to rewrite the array to preserve list semantics. In theory it is possible to iterate backwards though =)